src/app/clusters/operational-credentials-server/operational-credentials-server.cpp

```diff
@@ -314,15 +315,123 @@ bool emberAfOperationalCredentialsClusterAddOpCertCallback(chip::app::Command *
+    EmberAfStatus status   = EMBER_ZCL_STATUS_SUCCESS;
+    EmberStatus sendStatus = EMBER_SUCCESS;
+    CHIP_ERROR err         = CHIP_NO_ERROR;
+
+    chip::Platform::ScopedMemoryBuffer<uint8_t> csr;
+    size_t csrLength = Crypto::kMAX_CSR_Length;
+
+    uint8_t * csrBuf = nullptr;
+
+    constexpr uint16_t zclOctetStringHeaderSize = 1;
+
+    emberAfPrintln(EMBER_AF_PRINT_DEBUG, "OpCreds: commissioner has requested an OpCSR");
+
+    // Fetch current admin
+    AdminPairingInfo * admin = retrieveCurrentAdmin();
+    VerifyOrExit(admin != nullptr, status = EMBER_ZCL_STATUS_FAILURE);
+
+    // 1 extra byte for length in ZCL data type
+    VerifyOrExit(csr.Alloc(Crypto::kMAX_CSR_Length + zclOctetStringHeaderSize), status = EMBER_ZCL_STATUS_FAILURE);
+
+    if (admin->GetOperationalKey() == nullptr)
+    {
+        Crypto::P256Keypair keypair;
+        keypair.Initialize();
+        VerifyOrExit(admin->SetOperationalKey(keypair) == CHIP_NO_ERROR, status = EMBER_ZCL_STATUS_FAILURE);
+    }
+
+    csrBuf = csr.Get();
+    err    = admin->GetOperationalKey()->NewCertificateSigningRequest(&csrBuf[zclOctetStringHeaderSize], csrLength);
+    emberAfPrintln(EMBER_AF_PRINT_DEBUG, "OpCreds: NewCertificateSigningRequest returned %d", err);
+    VerifyOrExit(err == CHIP_NO_ERROR, status = EMBER_ZCL_STATUS_FAILURE);
+    VerifyOrExit(csrLength <= UINT8_MAX, status = EMBER_ZCL_STATUS_FAILURE);
+
+    // Fill in the length as the first byte of the octet string
+    csrBuf[0] = (uint8_t) csrLength;
+
+    if (commandObj == nullptr)
+    {
+        uint16_t nullValue = 0;
+        emberAfFillExternalBuffer((ZCL_CLUSTER_SPECIFIC_COMMAND | ZCL_FRAME_CONTROL_SERVER_TO_CLIENT),
+                                  ZCL_OPERATIONAL_CREDENTIALS_CLUSTER_ID, ZCL_OP_CSR_RESPONSE_COMMAND_ID, "bbbbbb", csr.Get(),
+                                  csrLength + zclOctetStringHeaderSize, CSRNonce.data(), CSRNonce.size(), &nullValue,
+                                  sizeof(nullValue), &nullValue, sizeof(nullValue), &nullValue, sizeof(nullValue), &nullValue,
+                                  sizeof(nullValue));
+        sendStatus = emberAfSendResponse();
+    }
+    else
+    {
+        app::CommandPathParams cmdParams = { emberAfCurrentEndpoint(), /* group id */ 0, ZCL_OPERATIONAL_CREDENTIALS_CLUSTER_ID,
+                                             ZCL_OP_CSR_RESPONSE_COMMAND_ID, (chip::app::CommandPathFlags::kEndpointIdValid) };
+
+        TLV::TLVWriter * writer = nullptr;
+
+        VerifyOrExit(commandObj != nullptr, err = CHIP_ERROR_INCORRECT_STATE);
+
+        SuccessOrExit(err = commandObj->PrepareCommand(&cmdParams));
+        writer = commandObj->GetCommandDataElementTLVWriter();
+        SuccessOrExit(err = writer->Put(TLV::ContextTag(0), ByteSpan(&csrBuf[1], csrLength)));
+        SuccessOrExit(err = writer->Put(TLV::ContextTag(1), CSRNonce));
+        SuccessOrExit(err = writer->Put(TLV::ContextTag(2), ByteSpan(nullptr, 0)));
```

Comment:
So this is sending an empty octet string.  But the emberAfFillExternalBuffer version (which I am guessing is not actually being exercised because commands are using the IM encoding?) sends... something really weird: it's two bytes of 0, and I am not sure what the other side is expected to do with that.

Reply:
Yeah, emberAfFillExternalBuffer version is no longer being used. The code was being written while we switched over from `emberAfFillExternalBuffer` to IM send version. Maybe we can just delete this code.